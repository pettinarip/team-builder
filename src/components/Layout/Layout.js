import React from 'react'
import PropTypes from 'prop-types'
import PlayerDroppable from '../PlayerDroppable'

const Layout = ({ config, playersPositions, addPlayer, removePlayer }) => {
  let index = 0
  const players = config.map(lineCount => {
    const slice = Array.apply(null, Array(lineCount)).map(
      (x, i) => ({
        ...playersPositions[index + i],
        position: index + i
      })
    )
    index += lineCount
    return slice
  })

  return (
    <div className='layout'>
      {players.reverse().map((line, i) =>
        <div className='line' key={i}>
          {line.map(player =>
            <PlayerDroppable
              className={player.id ? 'active' : ''}
              id={player.id}
              color={player.color}
              number={player.number}
              position={player.position}
              key={player.position}
              addPlayer={addPlayer}
              onClick={() => player.id && removePlayer(player.id, player.position)}
            />
          )}
        </div>
      )}
    </div>
  )
}

Layout.propTypes = {
  config: PropTypes.array,
  playersPositions: PropTypes.object,
  addPlayer: PropTypes.func.isRequired,
  removePlayer: PropTypes.func.isRequired
}

export default Layout
