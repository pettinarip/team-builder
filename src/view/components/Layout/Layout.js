import React,{ Component } from 'react'
import PropTypes from 'prop-types'
import { findDOMNode } from 'react-dom'
import Player from '../Player'
import Position from '../Position'

import ItemTypes from 'core/constants/ItemTypes'

import { DropTarget } from 'react-dnd'

class Layout extends Component {
  render () {
    const { config, playersPositions, connectDropTarget, switchPosition } = this.props
    const players = config.map((position, i) => ({
        ...playersPositions[i],
        ...position
      })
    )

    return connectDropTarget(
      <div className='layout'>
        {players.map((player, i) => (
            <Position
              id={i}
              key={i}
              x={player.x}
              y={player.y}
              onDropped={(source, target) => {
                if (typeof target !== 'undefined') {
                  switchPosition(source, target)
                }
              }}
            >
              <Player
                id={player.id}
                className={player.id ? 'active' : ''}
                number={player.number}
                key={player.position}
              />
            </Position>
          )
        )}
      </div>
    )
  }
}

Layout.propTypes = {
  config: PropTypes.array,
  playersPositions: PropTypes.object,
  updatePosition: PropTypes.func.isRequired,
  switchPosition: PropTypes.func.isRequired
}

export default DropTarget(ItemTypes.POSITION, {
  drop (props, monitor, component) {
    if (!monitor.didDrop()) {
      const dragItem = monitor.getItem()
      const fieldBoundingRect = findDOMNode(component).getBoundingClientRect()
      const sourceOffset = monitor.getSourceClientOffset()

      const x = sourceOffset.x - fieldBoundingRect.left
      const y = sourceOffset.y - fieldBoundingRect.top

      props.updatePosition(
        dragItem.position,
        x / fieldBoundingRect.width * 100,
        y / fieldBoundingRect.height * 100
      )
    }
  }
},
function (connect, monitor) {
  return {
    connectDropTarget: connect.dropTarget(),
    isOver: monitor.isOver()
  }
})(Layout)
