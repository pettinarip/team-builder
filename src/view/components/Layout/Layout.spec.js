import React from 'react'
import { shallow } from 'enzyme'
import Layout from './Layout'

describe('Layout component', () => {
  it('should render the component properly', () => {
    const config = [
      { x: 10, y: 10 },
      { x: 10, y: 20 },
      { x: 10, y: 30 }
    ]
    const component = mountLayout({
      config,
      playersPositions: {1: 1}
    })

    expect(component).toMatchSnapshot()
  })
})

function mountLayout (props) {
  const propsToUse = {
    updatePosition () {},
    switchPosition () {},
    config: {},
    playersPositions: {},
    ...props
  }

  const OriginalLayout = Layout.DecoratedComponent
  const identity = el => el;

  return shallow(<OriginalLayout {...propsToUse} connectDropTarget={identity} />)
}
