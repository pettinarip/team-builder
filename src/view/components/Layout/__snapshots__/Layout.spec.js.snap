// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout component should render the component properly 1`] = `
<div
  className="layout"
>
  <div
    className="line"
  >
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={8}
      removePlayer={[Function]}
    />
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={9}
      removePlayer={[Function]}
    />
  </div>
  <div
    className="line"
  >
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={4}
      removePlayer={[Function]}
    />
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={5}
      removePlayer={[Function]}
    />
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={6}
      removePlayer={[Function]}
    />
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={7}
      removePlayer={[Function]}
    />
  </div>
  <div
    className="line"
  >
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={0}
      removePlayer={[Function]}
    />
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={1}
      removePlayer={[Function]}
    />
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={2}
      removePlayer={[Function]}
    />
    <DropTarget(PlayerDroppable)
      addPlayer={[Function]}
      className=""
      position={3}
      removePlayer={[Function]}
    />
  </div>
</div>
`;
